File: examples/scara/sim_dba.py
import numpy as np
from scipy.interpolate import interp1d
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import matplotlib.colors as mcolors
import sys
from os.path import dirname, realpath
pypath = dirname(dirname(dirname(realpath(__file__)))) + '/python/'
sys.path.insert(1, pypath)
import utils
from odes import scara_2dbint, scara


# # Read the specification file
dirpath = dirname(realpath(__file__))
spec = "gb2"
specfile = '/' + spec + ".txt"
dba = utils.read_spec_from_txt(dirpath+specfile)


# # Read the controller files
tag = []
pavings = []
ctlr = []
for k in range(dba.n_dba):
    w = "/controller_" + spec + "_w" + str(k) + ".h5"
    tau, X, U, _, _, p, t, c = utils.read_controller_itvl_from_h5(dirpath+w)
    tag.append(t)
    pavings.append(p)
    ctlr.append(c)
controller = utils.CtlrItvl(U, pavings, tag, ctlr)
model = scara(tau)


# # Compute the percentage of winning set on the state space
winset = pavings[dba.q0][np.argwhere(tag[dba.q0] == 1).squeeze()]
print("\nWinning set coverage:")
wsize = np.sum((winset[:, 1]-winset[:, 0])
               * (winset[:, 3]-winset[:, 2])
               * (winset[:, 5]-winset[:, 4])
               * (winset[:, 7]-winset[:, 6]))
winper = "{:.2%}".format(
    wsize/((X[0, 1]-X[0, 0])*(X[1, 1]-X[1, 0])
           * (X[2, 1]-X[2, 0])*(X[3, 1]-X[3, 0]))
)
print(winper)


# # Set up workspace
nG = 2
G = np.zeros(shape=(4, 2, 2))
G[:, :, 0] = np.array([[0.4980, 0.5772], [1.5739, 1.7055],
                       [-0.1, 0.1], [-0.1, 0.1]])
G[:, :, 1] = np.array([[0.4903, 0.6069], [-0.9363, -0.8363],
                       [-0.1, 0.1], [-0.1, 0.1]])

A = pavings[dba.q0][np.argwhere(tag[dba.q0] == -1).squeeze(), :]


def get_labels(x, G, A):
    if(x[0] > G[0, 0, 0] and x[0] < G[0, 1, 0] and
       x[1] > G[1, 0, 0] and x[1] < G[1, 1, 0]):
        return 1
    elif(x[0] > G[0, 0, 1] and x[0] < G[0, 1, 1] and
         x[1] > G[1, 0, 1] and x[1] < G[1, 1, 1]):
        return 2
    elif(utils.index_in_interval_array(x, A).size > 0):
        return -1
    else:
        return 0


# # Simulation
rng = np.random.default_rng()
Tsim = 50

x0 = np.array([0.0, 0.0, 0.0, 0.0])
t = 0
x = x0
q = dba.q0
tsim = []
xsim = []
usim = []
qsim = []
torqsim = []
lsim = []
while(t < Tsim):
    x_id = utils.index_in_interval_array(x, pavings[q])
    if(x_id < 0):
        print("System state ")
        print(x)
        print(" is not inside the winning set.")
        break
    if(q < 0):
        print("System unsafe.")
        break

    if(any(ctlr[q][x_id, :])):
        uset = np.argwhere(ctlr[q][x_id, :]).squeeze()  # get the indices of valid input
    else:
        print("No valid control input.")
        break

    if(uset.size > 1):
        uid = rng.choice(uset, 1)  # randomly pick one
    else:
        uid = int(uset)
    u = U[uid, :].squeeze()

    # Calculate torque
    torq = model.compute_torque(np.atleast_2d(u).T, np.atleast_2d(x[2:4]).T,
                                np.atleast_2d(x[0:2]).T)

    # Integrate ode
    sol = solve_ivp(scara_2dbint, [0, tau], x, method='RK45', args=(u,))
    tt = sol.t[-1]
    y = sol.y[:, -1]

    # Save trajectories
    tsim.append(t)
    xsim.append(x)
    usim.append(u)
    qsim.append(q)
    torqsim.append(torq.astype(float).T.squeeze())
    lsim.append(get_labels(x, G, A))
    # Update state
    q = dba.q_prime[q, get_labels(x, G, A)]
    x = y
    t += tt

xsim = np.asarray(xsim)
usim = np.asarray(usim)
qsim = np.asarray(qsim)
tsim = np.asarray(tsim)
torqsim = np.asarray(torqsim)
lsim = np.asarray(lsim)


# # Save data to a .txt file
np.save(dirpath+'/traj_gb2.npy', xsim)
np.save(dirpath+'/torq_gb2.npy', torqsim)


# # x-y 2D plot of winning set
fig, ax = plt.subplots()
ax.set_xlim(X[0, 0], X[0, 1])
ax.set_xlabel(r'$\theta_1$')
ax.set_ylim(X[1, 0], X[1, 1])
ax.set_ylabel(r'$\theta_2$')

for i in range(nG):
    ax.add_patch(
        patches.Rectangle((G[0, 0, i], G[1, 0, i]),
                          G[0, 1, i]-G[0, 0, i], G[1, 1, i]-G[1, 0, i],
                          linewidth=1.5, edgecolor='g', fill=False)
    )
ax.add_collection(
    utils.polycoll_interval_array(A[:, 0:4], True, 'k', 0.7)
)
ax.add_collection(
    utils.polycoll_interval_array(winset[:, 0:4], True, 'palegoldenrod', 0.7)
)


# # Plot closed-loop trajectoryies
colors = list(mcolors.TABLEAU_COLORS.keys())
qdiff = np.diff(np.insert(qsim, 0, dba.q0))
qchks = np.argwhere(qdiff != 0).squeeze()
qchks = np.insert(qchks, 0, 0)
qchks = np.append(qchks, qsim.size)
q = dba.q0
for k in range(qchks.size-1):
    q = q + qdiff[qchks[k]]
    xs = xsim[qchks[k]:qchks[k+1]+1, :]
    ax.plot(xs[:, 0], xs[:, 1], color=colors[q])

ax.plot(xsim[0, 0], xsim[0, 1], marker='^',
        markerfacecolor='r', markeredgecolor='r')
ax.plot(xsim[-1, 0], xsim[-1, 1], marker='v',
        markerfacecolor='g', markeredgecolor='g')


figs, axs = plt.subplots(2)
t_interp = np.linspace(0, tsim[-1], 200)
fu = interp1d(tsim, torqsim, axis=0, kind='previous')
torq_interp = fu(t_interp)
fq = interp1d(tsim, qsim, axis=0, kind='previous')
q_interp = fq(t_interp)

axs[0].plot(t_interp, torq_interp)
axs[0].title.set_text('Time-Control Curves')
axs[0].legend((r'$\tau_1$', r'$\tau_2$'), loc="upper right")
axs[0].set_xlabel(r'$t$(s)')
axs[0].set_ylabel(r'$\tau_{1,2}$(N$\cdot$m)')
axs[1].plot(t_interp, q_interp)
axs[1].title.set_text('Time-Automaton State Curve')
axs[1].set_xlabel(r'$t$(s)')
axs[1].set_ylabel(r'$q$')
plt.subplots_adjust(left=None, bottom=None, right=None, top=None,
                    wspace=None, hspace=0.5)

plt.show()

File: examples/scara/sim_dba.m
clear
clc
addpath('../../matlab')
%% load spec & controller
fm= @twoIntegrators;
vf= @scara;

G1= [0.4980, 0.5772; 1.5739, 1.7055; -0.1 0.1; -0.1, 0.1];
G2= [0.4903, 0.6069; -0.9363, -0.8363; -0.1 0.1; -0.1, 0.1];

%%% Read the specification %%%
% - n_dba: # of DBA nodes.
% - n_props: # of all propositions (2^AP)
% - q0: the initial node of the DBA.
% - acc: accepting nodes.
% - q_prime: the DBA transition matrix.
spec= 'gb2';
[n_dba,n_props,q0,acc,q_prime]=read_spec(strcat(spec,'.txt'));

%%% Controller data %%%
% - ts: Sampling time.
% - U : All input values.
% - X : Workarea.
% - A(can be empty): obstacles.
% - G(can be empty): Target area.
% - pavings: Tree-structrued controller.
% - tag: indicating if a cell is inside the winning set.
% - ctlr: all valid control inputs for each cell in pavings.
controller.ctlr= cell(n_dba,1);
controller.partitions= cell(n_dba,1);
controller.tags= cell(n_dba,1);
controller.obs= cell(n_dba,1);

%%% If controllers are stored in *.mat %%%
% for i=1:n_dba
%     ctlrfile = strcat('data_', spec, '_w', num2str(i-1),'.mat');
%     load(ctlrfile);
%     controller.ctlr{i}= ctlr;
%     controller.partitions{i}= pavings;
%     controller.tags{i}= tag;
%     controller.obs{i}= controller.partitions{i}(controller.tags{i}==-1,:);
%     figure
%     plot2_boxes([controller.obs{i}(:,1:2),controller.obs{i}(:,3:4)],...
%         [0.5,0.5,0.5], 'k', 1)
% end

%%% If controllers are stored in *.h5 %%%
for i=1:n_dba
    ctlrfile = strcat('controller_', spec, '_w', num2str(i-1),'.h5');
    controller.ctlr{i}= h5read(ctlrfile, '/ctlr')';
    controller.partitions{i}= h5read(ctlrfile, '/pavings')';
    controller.tags{i}= h5read(ctlrfile, '/tag');
    controller.obs{i}= controller.partitions{i}(controller.tags{i}==-1,:);
end
X= h5read(ctlrfile, '/X')';
U= h5read(ctlrfile, '/U')';
ts= h5read(ctlrfile, '/ts');


%% simulation
x0= [0;0;0;0];
T= 50;

x= x0;  % a column
t= 0;
% q= [0,1,2];
qid= q0+1;
nacc= 0;

tsim= [];
xsim= [];  % a row
usim= [];
torqsim= [];
qsim= [];
rng('shuffle')
while(t<T)
    %%% compute control input %%%
    if (qid < n_dba+1)
        par= controller.partitions{qid};
        ctlr= controller.ctlr{qid};
        xid= find(x(1)>=par(:,1) & x(1)<=par(:,2) & ...
            x(2)>=par(:,3) & x(2)<=par(:,4) & ...
            x(3)>=par(:,5) & x(3)<=par(:,6) & ...
            x(4)>=par(:,7) & x(4)<=par(:,8));
        uid= find(ctlr(xid(1),:));
        if (isempty(uid))
            s1= 'Invalid controller for q';
            s2= num2str(qid-1);
            s= strcat(s1,s2);
            error(s);
        else
            % select a random u from all valid control values
            pick=randperm(numel(uid));
            u= U(uid(pick(1)),:);
            %         % select the minimum value
            %         uall= U(uid,:);
            %         [val, ind]= min(abs(uall(:,1)));
            %         u= uall(ind,:);
            
            %         % select the first/last value
            %         u= U(uid(end),:);
        end
    end
    
    %%% Convert from the angular accelarations to torques %%%
    torqsim_inter= compute_torque(u', x(3:4), x(1:2));
    
    tsim= [tsim; t];
    xsim= [xsim; x'];
    usim= [usim; u];
    torqsim= [torqsim; torqsim_inter'];
    qsim= [qsim; qid-1];
    
    %%% Update the state of the DBA %%%
    p= labeling(x, G1, G2);
    qid= q_prime(qid, p+1);
    if(qid-1==acc)
        nacc= nacc+1;
    end
    
    %%% Update the state of the dyanmical system %%%
    %%% use discrete-time two double integrators
    xx= fm(ts,x',u);
    x= xx';
    t= t+ts;
    %%% use the full scara ode
%     [tt, xx]= ode45(@(t,x) vf(t,x,torq), [0, ts], x);
%     x= xx(end,:)';
%     t= t + tt(end,:);
end


%% display
% define color
pink = [255,182,193]/255;
gold = [1,0.84,0];
lightgold = [255,236,139]/255;
gray = [0.6,0.6,0.6];
lightblue = [176 226 255]/255;
orange= [0.8500 0.3250 0.0980];

FS= 16; % fontsize
LW= 1.5; % lineweight

% % display the obstacles and target
figure
hold on
plot2_boxes([controller.obs{q0+1}(:,1:2),controller.obs{q0+1}(:,3:4)], ...
    [0.5,0.5,0.5], 'k', 1)
% % reference obstacles
l1 = 0.15;
l2 = 0.15;
h = 0.8*l1;
r = 0.5*l1;
a1 = atan2(h,r);
a2= asin(h/l1);
N1= 50;
N2= 30;
theta11= 0:a2/N1:a2;
theta21= pi - theta11 - atan2(h-l1.*sin(theta11), l1.*cos(theta11)-r);
theta12= a2:(a1-a2)/N2:a1;
theta22= pi - theta12 + atan2(l1.*sin(theta12)-h, l1.*cos(theta12));
theta1= [theta11, theta12];
theta2= [theta21, theta22];
plot(theta1, theta2, 'LineWidth', 2)
% % display target area
rectangle('Position', [G1(1,1), G1(2,1), ...
    G1(1,2)-G1(1,1), G1(2,2)-G1(2,1)],...
    'EdgeColor','g','FaceColor','g')
rectangle('Position', [G2(1,1), G2(2,1), ...
    G2(1,2)-G2(1,1), G2(2,2)-G2(2,1)],...
    'EdgeColor','g','FaceColor','g')

% % display phase trajectories
% plot(xsim(:,1),xsim(:,2),...
%     'Marker','.','MarkerEdgeColor','r','MarkerFaceColor','r')
plot(xsim(:,1),xsim(:,2), 'LineWidth', LW)

axis([X(1,:) X(2,:)])
xlabel({'$\theta_1$ (rad)'}, 'interpreter','latex',...
    'Units','points', 'FontSize',FS,...
    'FontName','Times', 'FontWeight','bold')
ylabel({'$\theta_2$ (rad)'}, 'interpreter','latex',...
    'Units','points', 'FontSize',FS,...
    'FontName','Times', 'FontWeight','bold')

% % display computed torque
tq= [0:0.01:tsim(end)]';
torqsim_inter= interp1(tsim,torqsim,tq,'previous');

figure
plot(tq, torqsim_inter, 'LineWidth', LW)
axis([0, tsim(end), min(min(torqsim)), max(max(torqsim))])
xlabel({'$t(s)$'}, 'interpreter','latex',...
    'Units','points', 'FontSize',FS,...
    'FontName','Times', 'FontWeight','bold')
ylabel({'$\tau_1,\;\tau_2$'}, 'interpreter','latex',...
    'Units','points', 'FontSize',FS,...
    'FontName','Times', 'FontWeight','bold')

hl= legend({'$\tau_1(t)$','$\tau_2(t)$'}, 'Interpreter', 'latex',...
    'Units','points', 'FontSize',FS,...
    'FontName','Times', 'FontWeight','bold');